src/main/java/com/google/devtools/build/lib/syntax/MethodLibrary.java

```diff
@@ -170,17 +218,47 @@ private static boolean hasElementWithBooleanValue(
             type = Object.class,
             doc = "This collection.",
             // TODO(cparsons): This parameter should be positional-only.
-            legacyNamed = true)
+            legacyNamed = true),
+        @Param(
+            name = "key",
+            doc = "An optional function applied to each element before comparison.",
+            named = true,
+            defaultValue = "None",
+            positional = false,
+            noneable = true),
+        @Param(
+            name = "reverse",
+            type = Boolean.class,
+            doc = "Return results in descending order.",
+            named = true,
+            defaultValue = "False",
+            positional = false,
+            noneable = true)
       },
       useLocation = true,
       useEnvironment = true)
-  public MutableList<?> sorted(Object self, Location loc, Environment env) throws EvalException {
-    try {
-      return MutableList.copyOf(
-          env, EvalUtils.SKYLARK_COMPARATOR.sortedCopy(EvalUtils.toCollection(self, loc, env)));
-    } catch (EvalUtils.ComparisonException e) {
-      throw new EvalException(loc, e);
+  public MutableList<?> sorted(Object self, Object key, Boolean reverse,
+      Location loc, Environment env)
+      throws EvalException, InterruptedException {
+
+    List list = new ArrayList(EvalUtils.toCollection(self, loc, env));
+    KeyComparator comparator = new KeyComparator(key, loc, env);
```

Comment:
If key is None, we don't need to bother with KeyComparator; just use SKYLARK_COMPARATOR directly. Also, if key is not a SkylarkFunction, we should reject the argument here, before we reach sort.